experimental/JuLie/partitions.jl

```diff
@@ -0,0 +1,669 @@
+################################################################################
+# Partitions.
+#
+# Copyright (C) 2020 Ulrich Thiel, ulthiel.com/math
+#
+# Originally taken from the JuLie [repository](https://github.com/ulthiel/JuLie)
+# by Ulrich Thiel and OSCAR-ified by Claudia He Yun and Matthias Zach.
+################################################################################
+
+export Partition, partitions, ascending_partitions, dominates, conjugate, getindex_safe, num_partitions
+
+@doc Markdown.doc"""
+    Partition{T} <: AbstractVector{T}
+
+A **partition** of an integer ``n ≥ 0`` is a decreasing sequence ``λ=(λ₁,…,λᵣ)`` of positive integers ``λᵢ`` whose sum is equal to ``n``. The ``λᵢ`` are called the **parts** of the partition. We encode a partition as an array with elements ``λᵢ``. You may increase performance by using smaller integer types, see the examples below. For efficiency, the ```Partition``` constructor does not check whether the given array is in fact a partition, i.e. a decreasing sequence.
+
+# Examples
+```julia-repl
+julia> P=Partition([3,2,1]) #The partition 3+2+1 of 6
+[3, 2, 1]
+julia> sum(P) #The sum of the parts.
+6
+julia> P[1] #First component
+3
+julia> P=Partition(Int8[3,2,1]) #Same partition but using 8 bit integers
```

Comment:
All examples should be (if at all possible) using `jldoctest`, not `julia-repl`, to ensure they stay valid and are tested as part of our CI tests.
````suggestion
```jldoctest
julia> P=Partition([3,2,1]) #The partition 3+2+1 of 6
[3, 2, 1]

julia> sum(P) #The sum of the parts.
6

julia> P[1] #First component
3

julia> P=Partition(Int8[3,2,1]) #Same partition but using 8 bit integers
Int8[3, 2, 1]
````